DSpace/dspace-angular
# Fix themed components duplicating themself when switching themes

## References
* Closes #2346

## Description
This PR fixes 4 issues related to theming:
-  A bug where the old component is not destroyed when the theme is updated which leads to the every themeable component being duplicated every time the theme changes.
- A rare issue in production mode where the `@ViewChild` variables can sometimes still be uninitialised when trying to render the component, leading to the component not rendering.
- An error where the base theme wouldn't be restored after leaving a UUID/handle theme (this issue only happend when there was no default/fallback theme set).
- An infinite loop caused by the `createHeadTags` method when not overriding the `headTags` property in you themes (only the case when you have themes extending each other)

## Instructions for Reviewers
List of changes in this PR:
* Changed the `ngOnInit()` to `ngAfterViewInit()`, because `@ViewChild` is only initialised after the OnInit, but right before AfterViewInit (see [docs](https://angular.io/api/core/ViewChild)). This is often not an issue because Angular is slow enough because of relatively large/complex HTML templates, and by the time the `ThemedComponent` tries to create the component the AfterViewInit hook will often already have been called. But sometimes it does resolve fast enough (for example the `HomeNewsComponent` which only contains text and an image) and then the code breaks when it [tries to create the component](https://github.com/DSpace/dspace-angular/blob/404ccd9b0ecdcdfe63ad22779859b1a01eddc12e/src/app/shared/theme-support/themed.component.ts#L124) because `vcr` is still `undefined`.
* Ensured that `destroyComponentInstance()` is always triggered before the `ThemedComponent` creates it's new component.
* Ensured that the base theme is restored when leaving a UUID/handle theme, this was done by fixing the `ThemeService. getActionForMatch` method who doesn't update the current theme, when the new fallback/default theme is not set. This is for example the case when you only use this theme config:
  ```yml
  themes:
    - name: dspace
      handle: '123456789/1'
  ```

**Guidance for how to test/review this PR:**
Error 1:
- Add the following config to you `config.yml` file:
  ```yml
  themes:
    - name: dspace
      extends: custom
      handle: '123456789/1'
    - name: custom
  ```
- Go to the home page and check that the custom theme is being used
- Go to the top level community `123456789/1` and check that the dspace theme is being used (you can easily spot it by looking at the header)

Error 3:
- Replace the `themes` config with the following config in your `config.yml` file:
  ```yml
  themes:
    - name: dspace
      handle: '123456789/1'
  ```
- Go to the top level community `123456789/1` and check that the dspace theme is being used (you can easily spot it by looking at the header)
- Go back to the home page and check that the base theme is used again

Error 4:
- Use the following config:
   ```yml
   themes:
     - name: dspace
        extends: custom
     - name: custom
   ```
- Check that the `createHeadTags` method isn't throwing an infinite loop error anymore in the browser console
- Check that the default favicon is used
- Overwrite the `headTags` in the custom theme by adding a custom favicon for example & check that the new favicon is used
- Overwrite the `headTags` in the dspace theme by adding a custom favicon for example & check that that new favicon is used instead of the one from the custom theme

## Checklist
